src/managers/channelManager/shared/SecondaryChannelEmail.ts

```diff
@@ -0,0 +1,111 @@
+import { NotSubscribedError, NotSubscribedReason } from "../../../errors/NotSubscribedError";
+import Log from "../../../libraries/Log";
+import { EmailProfile } from "../../../models/EmailProfile";
+import { SecondaryChannelProfile } from "../../../models/SecondaryChannelProfile";
+import OneSignalApi from "../../../OneSignalApi";
+import Database from "../../../services/Database";
+import { SecondaryChannel, SecondaryChannelWithControllerEvents } from "./SecondaryChannel";
+import { SecondaryChannelController } from "./SecondaryChannelController";
+import { SecondaryChannelIdentifierUpdater } from "./updaters/SecondaryChannelIdentifierUpdater";
+import { SecondaryChannelTagsUpdater } from "./updaters/SecondaryChannelTagsUpdater";
+import { SecondaryChannelExternalUserIdUpdater } from "./updaters/SecondaryChannelExternalUserIdUpdater";
+import { SecondaryChannelFocusUpdater } from "./updaters/SecondaryChannelFocusUpdater";
+import { SecondaryChannelSessionUpdater } from "./updaters/SecondaryChannelSessionUpdater";
+
+export class SecondaryChannelEmail implements SecondaryChannel, SecondaryChannelWithControllerEvents {
+
+  constructor(
+    readonly secondaryChannelController: SecondaryChannelController,
+    readonly secondaryChannelIdentifierUpdater: SecondaryChannelIdentifierUpdater,
+    readonly secondaryChannelExternalUserIdUpdater: SecondaryChannelExternalUserIdUpdater,
+    readonly secondaryChannelTagsUpdater: SecondaryChannelTagsUpdater,
+    readonly secondaryChannelSessionUpdater: SecondaryChannelSessionUpdater,
+    readonly secondaryChannelFocusUpdater: SecondaryChannelFocusUpdater,
+    ) {
+    secondaryChannelController.registerChannel(this);
+  }
+
+  async logout(): Promise<boolean> {
+    const { deviceId } = await Database.getSubscription();
+    if (!deviceId) {
+      Log.warn(new NotSubscribedError(NotSubscribedReason.NoDeviceId));
+      return false;
+    }
+
+    const emailProfile = await Database.getEmailProfile();
+    if (!emailProfile.playerId) {
+      Log.warn(new NotSubscribedError(NotSubscribedReason.NoEmailSet));
+      return false;
+    }
+
+    const appConfig = await Database.getAppConfig();
+
+    if (!await OneSignalApi.logoutEmail(appConfig, emailProfile, deviceId)) {
+      Log.warn("Failed to logout email.");
+      return false;
+    }
+
+    await Database.setEmailProfile(new EmailProfile());
+    return true;
+  }
+
+  async setIdentifier(identifier: string, authHash?: string): Promise<string | null> {
+    const profileProvider = this.secondaryChannelIdentifierUpdater.profileProvider;
+    const existingEmailProfile = await profileProvider.getProfile();
```

Comment:
I'm a bit confused here. Is it possible that there is no email profile yet?

If yes, maybe change to `possibleExistingEmailProfile` or something similar

Reply:
Yep that works